### DISCUSSION TITLE
Second Factor Authentication?

### DISCUSSION THREAD

**Discussion by akatsev** [IbcAlpha/IBC#259] [Q&A] [answered]
Is there any way to bypass/suppress/override/cancel Second Factor Authentication?

**Comment by rlktradewright** [comment]
No, for live acounts.

For paper-trading accounts, 2FA is not used at all.

But really, it could hardly be any simpler. Once a week (assuming you use auto-restart), you have to make sure your phone is turned on, and then acknowledge the alert. Why is that a problem?

**Reply by akatsev** [reply]
Sometimes I get error codes 1100 or 2110, which makes me assume that something is wrong with the TWS/IBG, and then I restart it.
Another situation is restarting the computer itself due to a Windows update or power failure or somewhat else...

**Comment by rlktradewright** [comment]
You need to understand the meaning of these error codes. Read the IB documentation, study, ask questions on the forum if you don't understand. Neither of those errors is a reason to restart TWS/Gateway. In fact there is only very rarely any reason to do so.

If you're getting WIndows updates during the week, check out the controls to see what you can do. For example, each Saturday I check for updates, and if there are any I then pause updates for 7 days (after applying them). This means I only ever get updates at the weekend.

If you're running the Home version of WIndows, these controls may be less flexible (I don't know because I don't use it). If so, upgrade to the Professional Edition. Or buy Windows Server, which gives you pretty much complete control over updates. Or run TWS/Gateway on Linux.

If you suffer from frequent power failures, buy a UPS, or get a generator.

**Reply by akatsev** [reply]
Thank you very much. Your explanation regarding Windows updates etc are very helpful.
Regarding 1100 and 2110. I've read their explanation in "Message Codes" paragraph of IB API documentation and I find them insufficient in order to decide regarding the need of TWS/IBG restart. Then I asked IB support chat and get the clear answer, that in the case of 1100 it is necessary to restart TWS/IBG.
I'd like to get detailed and propper explanations about these errors, but I don't know, where can I find them.

**Comment by rlktradewright** [comment, marked as answer]
For questions about the TWS API, your should ask at the TWS API Users Group at https://groups.io/g/twsapi

Error 1100 means that TWS/GAteway has lost its connection to the IB servers (you are not told the reason). When that happens, TWS/Gateway tries to reconnect, and when that succeeds, you get an 1101 or 1102 'error' which tells you that the connection has been re-established and whether any data has been lost. So when you get 1101 or 1102, your application can reconnect to the API and continue. There is absolutely no reason to restart TWS/Gateway. So either IB have misled you, or they didn't really understand what you were asking.

By the way, error 1100 occurs every day during the nightly reset, and the reconnection is typically a minute or two later. It may also occur occasionally during the day, and then the reconnection is usually after a few seconds. Longer disconnections are rare.

This is not the right place for discussing the API, so please join the Group mentioned above and ask there if you need further help.

**Reply by akatsev** [reply]
Great! Thank you so much!

**Comment by XYUU** [comment]
I have implemented the automated two‑factor authentication feature. This was added because the need for it became clear, and the idea was inspired by previous discussions. TOTP-based authentication is optional rather than mandatory. As a user, you may choose full automation or continue confirming authentication through IB Key manually. Once automation is enabled, you are responsible for the security of your own device and environment.

I do not assume any responsibility for losses or damages resulting from the use of this feature. It is open‑source and entirely optional.

https://github.com/XYUU/IBC

**Reply by rlktradewright** [reply]
I won't endorse any IBC replacement that automates TOTP (or any other 2FA method).

Automating 2FA bypasses the security mechanisms that IBKR have introduced, which not only provide additional security to users but also provide evidence to IBKR that the user is indeed the intended user, which would be important in case of claims for loss. Automation completely negates this, especially if it doesn't involve a user at all (such as where username and password entry is also automated), and I believe that IBKR would take a very dim view of this. Anyone who wants to follow this path would be advised to take legal advice.

**Reply by XYUU** [reply]
Thanks for pointing out the potential risks. I honestly hadn’t considered the legal implications at the time. I’ve already deleted my fork and repository — I definitely don’t want to get myself into any trouble because of this.
